sns-testing: SNS lifecycle

rvem opened the pull request.

rvem committed:
sns-testing: propose, adopt, and complete swap for the new SNS

1) Add a simple test canister that will be decentralized as a part of
   newly created SNS
2) Add helpers to install this canister and run it as the new SNS
3) Complete the SNS swap to finalize SNS adoption

rvem committed:
sns-testing: upgrade the test canister

Add helpers to upgrade the test canister.
Currently, the helper will reuse the same WASM module, but it's possible
to provide the new state value as a 'post_upgrade' hook argument.

Test scenario as well as CLI now perform canister upgrade. The former
also allows to see the SNS voting procedure in the NNS dapp web UI.

rvem commented on rs/nervous_system/integration_tests/src/pocket_ic_helpers.rs:
```diff
@@ -390,25 +396,32 @@ pub async fn install_all_nns_canisters(
     pocket_ic: &PocketIc,
     initial_balances: Vec<(AccountIdentifier, Tokens)>,
     with_mainnet_nns_canister_versions: bool,
+    with_test_nns_governance_canister: bool,
```
TODO: refactor once https://github.com/dfinity/ic/pull/3666 is merged in the upstream

sras commented on rs/sns/testing/src/pocket_ic.rs:
```diff
@@ -1,20 +1,38 @@
+            },
+            swap::{await_swap_lifecycle, smoke_test_participate_and_finalize},
+        },
+    },
```
Seems like nesting is a bit too much here in these imports.

sras commented on rs/nervous_system/integration_tests/src/pocket_ic_helpers.rs:
```diff
@@ -390,25 +396,32 @@ pub async fn install_all_nns_canisters(
-/// 5. `install_index_canister` - a flag indicating whether the Index canister for ICP Ledger should be installed.
-/// 6. `install_cmc` - a flag indicating whether the Cycles minting canister should be installed.
+/// 4. `custom_initial_registry_mutations` are custom mutations for the inital Registry. These
+///    should mutations should comply with Registry invariants, otherwise this function will fail.
```
it seems that "should" is redundant

rvem replied:
FWIW, this typo was there before my changes, but I'll fix it

sras commented on rs/nervous_system/integration_tests/src/pocket_ic_helpers.rs:
```diff
@@ -390,25 +396,32 @@ pub async fn install_all_nns_canisters(
 ) -> Vec<PrincipalId> {
+    assert!(
+        !(with_mainnet_nns_canister_versions && with_test_nns_governance_canister),
+        "Cannot mainnet canister version together wit NNS governance test canister version"
```
>   "Cannot mainnet canister version together wit NNS governance test canister version"

Did you mean "Cannot use mainnet" ? 

>wit

Typo?

sras commented on rs/sns/testing/src/pocket_ic.rs:
```diff
@@ -124,3 +142,96 @@ async fn install_frontend_nns_canisters(pocket_ic: &PocketIc) {
+    let (sns, proposal_id) =
+        propose_to_deploy_sns_and_wait(pocket_ic, create_service_nervous_system, sns_proposal_id)
+            .await;
+    await_swap_lifecycle(pocket_ic, sns.swap.canister_id, Lifecycle::Open)
```
The issue seem to require this

> sns-testing scenario should close the swap using a sufficient amount of test participants with a sufficient amount of ICPs.

Do we know that it works like this in this `await_swap_lifecycle`? Or is it done somewhere else?

rvem replied:
> Or is it done somewhere else?

The swap is handled in `smoke_test_participate_and_finalize` slightly below. This helper considers required parameters in order to correctly close the swap

rvem committed:
nervous_system::integration_tests: live mode pocket-ic helpers

Problem: Currently, helpers from 'pocket_ic_helpers' module use
'advance_time' pocket-ic method to artificially advance time in order to
wait for certain events to happen.
However, using 'advance_time' in live mode breaks agent certificates
checking since pocket-ic time becomes larger than the real time.

Solution:
1) When pocket-ic is running in live mode, run 'std::thread::sleep()'
   instead of 'advance_time()'.
2) Avoid waiting two days for SNS swap to start in live mode.
3) Allow to use the test version of NNS Governance
   canister which is capable of starting SNS swap immediately after SNS
   proposal adoption.

rvem committed:
sns-testing: use NNS Governance test version

Problem: We want to be able to run pocket-ic in live mode. However, this
is not viable with non-test version of NNS Governance canister because
it postpones the SNS swap start.

Solution: Use test version of NNS Governance canister. Ensure that swap
start_time is 'None' to start the swap immediately after SNS creation
proposal adoption.

sras commented on rs/sns/testing/canister/canister.rs:
```diff
@@ -0,0 +1,44 @@
+}
+
+fn init_impl(x: Option<InitArgs>) {
+    match x {
+        None => (),
+        Some(x) => {
+            match x.greeting {
+                None => (),
+                Some(g) => {
+                    STR.with(|s| *s.borrow_mut() = g);
+                }
+            };
+        }
+    }
```
May be using `if let` is better here..

sras reviewed (approved):
LGTM
